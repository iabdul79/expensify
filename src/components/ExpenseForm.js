import React from 'react'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props)
    const {
      description = '',
      note = '',
      createdAt = undefined,
      amount = ''
    } = props.expense ?? {}
    this.state = {
      description,
      note,
      createdAt: moment(createdAt),
      amount: parseFloat(amount / 100).toString(),
      calenderFocused: false,
      error: ''
    }
  }

  componentDidUpdate(prevProps) {
    if (this.props.expense && prevProps.expense?.id !== this.props.expense.id) {
      this.setState(() => {
        const {
          description,
          note,
          createdAt,
          amount
        } = this.props.expense
        return {
          description,
          note,
          createdAt: moment(createdAt),
          amount: parseFloat(amount / 100).toString(),
          calenderFocused: false,
          error: ''
        }
      })
    }
  }

  onDescriptionChange = (e) => {
    const description = e.target.value
    this.setState(() => ({ description }))
  }

  onNoteChange = (e) => {
    const note = e.target.value
    this.setState(() => ({ note }))
  }

  onAmountChange = (e) => {
    const amount = e.target.value
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }))
    }
  }

  onDateChanged = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }))
    }
  }

  onFocusChange = ({ focused }) => {
    this.setState(() => ({calenderFocused: focused}))
  }

  onSubmit = (e) => {
    e.preventDefault()
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({error: 'Please provide description and amount.'}))
      return
    }
    this.setState(() => ({error: ''}))
    this.props.onSubmit({
      description: this.state.description,
      amount: parseFloat(this.state.amount, 10) * 100,
      createdAt: this.state.createdAt.valueOf(),
      note: this.state.note
    })
  }

  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <h4 className="form__error"><i>{this.state.error}</i></h4>}
        <input
          type="text"
          className="input-text"
          placeholder="Description"
          autoFocus
          value={this.state.description}
          onChange={this.onDescriptionChange}
        />
        <input
          type="text"
          className="input-text"
          placeholder="Amount"
          value={this.state.amount}
          onChange={this.onAmountChange}
        />
        <SingleDatePicker 
          date={this.state.createdAt}
          onDateChange={this.onDateChanged}
          focused={this.state.calenderFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          displayFormat={'DD MMM YYYY'}
        />
        <textarea
          className="input-textarea"
          placeholder="Add a note for your expense (optional)"
          value={this.state.note}
          onChange={this.onNoteChange}
        />
        <div>
          <button className="button">Add expense</button>
        </div>
      </form>
    )
  }
}