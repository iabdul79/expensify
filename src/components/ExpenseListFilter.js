import React from 'react'
import { connect } from 'react-redux'
import { setEndDate, setStartDate, setTextFilter, sortByAmount, sortByDate } from '../actions/filters-actions'
import { DateRangePicker } from 'react-dates'

export class ExpenseListFilter extends React.Component {
  constructor(props) {
    super(props)
    this.state = {}
  }

  onFilterTextChange = (e) => {
    this.props.setTextFilter(e.target.value)
  }

  onSortByChange = (e) => {
    const sortBy = e.target.value
    const action = sortBy === 'amount' ? 'sortByAmount' : 'sortByDate'
    this.props[action]()
  }

  onDatesChanged = ({startDate, endDate}) => {
    this.props.setStartDate(startDate)
    this.props.setEndDate(endDate)
  }

  onFocusChange = (calenderFocused) => {
    this.setState(() => ({calenderFocused}))
  }

  render() {
    return (
      <div className="main-container">
        <div className="input-group">
          <div className="input-group__item">
            <input type="text" className="input-text" placeholder="Search expense" value={this.props.filters.text} onChange={this.onFilterTextChange}/>
          </div>
          <div className="input-group__item">
            <select className="input-select" value={this.props.filters.sortBy} onChange={this.onSortByChange}>
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker
              startDate={this.props.filters.startDate}
              endDate={this.props.filters.endDate}
              onDatesChange={this.onDatesChanged}
              focusedInput={this.state.calenderFocused}
              onFocusChange={this.onFocusChange}
              showClearDates={true}
              numberOfMonths={1}
              isOutsideRange={() => false}
              displayFormat={'DD MMM YYYY'}
            />
          </div>
        </div>
      </div>
    )
  }
}

const stateMapper = (state) => ({
  filters: state.filters
})

const mapDispatchToProps = (dispatch) => (
  {
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByAmount: () => dispatch(sortByAmount()),
    sortByDate: () => dispatch(sortByDate()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
  }
)

const connector = connect(stateMapper, mapDispatchToProps)

export default connector(ExpenseListFilter)