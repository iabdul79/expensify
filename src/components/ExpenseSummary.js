import React from 'react'
import numeral from 'numeral'
import { getVisibleExpense } from '../selectors/expense-selectors'
import { getExpenseTotal } from '../selectors/expense-total'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'

export class ExpenseSummary extends React.Component {
  render() {
    return (
      <div className="page-header">
        <div className="main-container">
          <h1 className="page-header__title">
            Viewing <span>{this.props.count}</span> expenses totalling <span>{this.props.total}</span>
          </h1>
          <div className="page-header__action">
            <Link className="button" to="/create">Add Expense</Link>
          </div>
        </div>
      </div>
    )
  }
}

const mapStateToProps = (state) => {
  const expenses = getVisibleExpense(state.expenses, state.filters)
  return {
    count: expenses.length,
    total: numeral(getExpenseTotal(expenses)).format('$0,0.00')
  }
}

const connector = connect(mapStateToProps)

export default connector(ExpenseSummary)