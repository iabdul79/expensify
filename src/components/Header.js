import React from 'react'
import { connect } from 'react-redux'
import {Link} from 'react-router-dom'
import { logoutRx } from '../actions/auth-actions'

export const Header = ({logOut}) => (
  <header className="header">
    <div className="main-container header__container">
      <Link className="header__title" to="/dashboard">
        <h1>Expensify</h1>
      </Link>
      <button className="button button--link" onClick={logOut}>Log Out</button>
    </div>
  </header>
)
const mapDispatchToProps = (dispatch) => ({
  logOut: () => dispatch(logoutRx())
})
export default connect(undefined, mapDispatchToProps)(Header)