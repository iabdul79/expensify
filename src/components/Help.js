import React from 'react'

export default () => (
  <p>This is help page</p>
)