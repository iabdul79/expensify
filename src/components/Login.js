import React from 'react'
import { connect } from 'react-redux'
import { loginRx } from '../actions/auth-actions'

export const Login = ({login}) => (
  <div className="box-layout">
    <div className="box-layout__box">
      <h1 className="box-layout__title">Expensify</h1>
      <p>Time to get your expense correct</p>
      <button className="button" onClick={login}>Login with Google</button>
    </div>
  </div>
)

const mapDispatchToProps = (dispatch) => (
  {
    login: () => dispatch(loginRx())
  }
)

export default connect(undefined, mapDispatchToProps)(Login)
