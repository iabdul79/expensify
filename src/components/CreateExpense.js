import React from 'react'
import ExpenseForm from './ExpenseForm'
import {connect} from 'react-redux'
import {addExpenseRx} from '../actions/expense-actions'


export class AddExpense extends React.Component {
  addExpense = (expense) => {
    this.props.addExpense(expense)
    this.props.history.push('/dashboard')
  }
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="main-container">
            <h1 className="page-header__title">Save Expense</h1>
          </div>
        </div>
        <div className="main-container">
          <ExpenseForm
            onSubmit={this.addExpense}
          />
        </div>
      </div>
    )
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    addExpense: (expense) => dispatch(addExpenseRx(expense))
  }
}

export default connect(undefined, mapDispatchToProps)(AddExpense)