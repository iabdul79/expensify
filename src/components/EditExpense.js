import React from 'react'
import { connect } from 'react-redux'
import { editExpense, removeExpense, removeExpenseRx, updateExpenseRx } from '../actions/expense-actions'
import ExpenseForm from './ExpenseForm'

export class EditExpense extends React.Component {
  editExpense = (updatedExpense) => {
    this.props.editExpense(this.props.expenseId, updatedExpense)
    this.props.history.push('/dashboard')
  }
  removeExpense = () => {
    this.props.removeExpense(this.props.expenseId)
    this.props.history.push('/dashboard')
  }

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="main-container">
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="main-container">
          <ExpenseForm
            expense={this.props.expense}
            onSubmit={this.editExpense}
          />
          <button className="button button--secondary" onClick={this.removeExpense}
          >Remove expense</button>
        </div>
      </div>
    )
  }
}

const mapStateToProp = (state, props) => (
  {
    expenseId: props.match.params.id,
    expense: state.expenses.find(expense => expense.id === props.match.params.id)
  }
)

const mapDispatchToProp = (dispatch) => (
  {
    editExpense: (expenseId, updateExpense) => dispatch(updateExpenseRx(expenseId, updateExpense)),
    removeExpense: (expenseId) => dispatch(removeExpenseRx(expenseId))
  }
)

const connector = connect(mapStateToProp, mapDispatchToProp)

export default connector(EditExpense)

