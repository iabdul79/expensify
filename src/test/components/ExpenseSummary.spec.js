import React from 'react'
import {shallow} from 'enzyme'
import {ExpenseSummary} from '../../components/ExpenseSummary'

test('should render summary component', () => {
  const wrapper = shallow(<ExpenseSummary count={3} total={1234} />)
  expect(wrapper).toMatchSnapshot()
})