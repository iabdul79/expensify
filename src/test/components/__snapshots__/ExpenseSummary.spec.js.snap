// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render summary component 1`] = `
<div
  className="page-header"
>
  <div
    className="main-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        3
      </span>
       expenses totalling 
      <span>
        1234
      </span>
    </h1>
    <div
      className="page-header__action"
    >
      <[object Object]
        className="button"
        to="/create"
      >
        Add Expense
      </[object Object]>
    </div>
  </div>
</div>
`;
